src/node/stage/comm.rs

```diff
@@ -6,136 +6,447 @@
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn successful_send_to_subset() -> Result<()> {
+        let comm = Comm::new(transport_config())?;
+
+        let mut peer0 = Peer::new()?;
+        let mut peer1 = Peer::new()?;
+
+        let message = Bytes::from_static(b"hello world");
+        comm.send_message_to_targets(&[peer0.addr, peer1.addr], 1, message.clone())
+            .await?;
+
+        assert_eq!(peer0.rx.recv().await, Some(message));
+
+        assert!(time::timeout(Duration::from_millis(100), peer1.rx.recv())
+            .await
+            .unwrap_or_default()
+            .is_none());
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn failed_send() -> Result<()> {
+        let comm = Comm::new(transport_config())?;
+        let invalid_addr = get_invalid_addr().await?;
+
+        let message = Bytes::from_static(b"hello world");
+        match comm
+            .send_message_to_targets(&[invalid_addr], 1, message.clone())
+            .await
+        {
+            Err(error) => assert_eq!(error.failed_recipients, [invalid_addr]),
+            Ok(_) => panic!("unexpected success"),
+        }
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn successful_send_after_failed_attempts() -> Result<()> {
+        let comm = Comm::new(transport_config())?;
+        let mut peer = Peer::new()?;
+        let invalid_addr = get_invalid_addr().await?;
+
+        let message = Bytes::from_static(b"hello world");
+        comm.send_message_to_targets(&[invalid_addr, peer.addr], 1, message.clone())
+            .await?;
+
+        assert_eq!(peer.rx.recv().await, Some(message));
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn partially_successful_send() -> Result<()> {
+        let comm = Comm::new(transport_config())?;
+        let mut peer = Peer::new()?;
+        let invalid_addr = get_invalid_addr().await?;
+
+        let message = Bytes::from_static(b"hello world");
+
+        match comm
+            .send_message_to_targets(&[invalid_addr, peer.addr], 2, message.clone())
             .await
+        {
+            Ok(_) => panic!("unexpected success"),
+            Err(error) => assert_eq!(error.failed_recipients, [invalid_addr]),
+        }
+
+        assert_eq!(peer.rx.recv().await, Some(message));
+
+        Ok(())
+    }
+
+    fn transport_config() -> qp2p::Config {
+        qp2p::Config {
+            ip: Some(IpAddr::V4(Ipv4Addr::LOCALHOST)),
+            idle_timeout_msec: Some(1),
+            ..Default::default()
+        }
+    }
+
+    struct Peer {
+        addr: SocketAddr,
+        rx: mpsc::Receiver<Bytes>,
+    }
+
+    impl Peer {
+        fn new() -> Result<Self> {
+            let transport = QuicP2p::with_config(Some(transport_config()), &[], false)?;
+
+            let endpoint = transport.new_endpoint()?;
+            let addr = endpoint.local_addr()?;
+            let mut incoming_connections = endpoint.listen()?;
+
+            let (tx, rx) = mpsc::channel(1);
```

Comment:
make this `1` as a const? 
maybe it shall be the same as the above idle_timeout_msec ?

Reply:
It's the channel capacity. It means it can buffer at most one item. Anything more, or even using unbounded channel here would be overkill. It has nothing to do with the idle timeout. Might move it a const, but doesn't seem it would bring us that much.